src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy

```diff
@@ -43,55 +114,52 @@ class ComposeUp extends DefaultTask {
         wasReconnected = false
-        if (settings.buildBeforeUp.get()) {
-            settings.buildTask.get().build()
-        }
         String[] args = ['up', '-d']
-        if (settings.removeOrphans()) {
+        if (composeExecutor.get().shouldRemoveOrphans()) {
             args += '--remove-orphans'
         }
-        if (settings.forceRecreate.get()) {
+        if (forceRecreate.get()) {
             args += '--force-recreate'
             args += '--renew-anon-volumes'
-        } else if (settings.noRecreate.get()) {
+        } else if (noRecreate.get()) {
             args += '--no-recreate'
         }
-        if (settings.scale()) {
-            args += settings.scale.get().collect { service, value ->
+        if (composeExecutor.get().isScaleSupported()) {
+            args += scale.get().collect { service, value ->
                 ['--scale', "$service=$value"]
             }.flatten()
         }
-        args += settings.upAdditionalArgs.get()
-        args += settings.startedServices.get()
+        args += upAdditionalArgs.get()
+        args += startedServices.get()
         try {
             def composeLog = null
-            if(settings.composeLogToFile.isPresent()) {
-              File logFile = settings.composeLogToFile.get().asFile
+            if (composeLogToFile.isPresent()) {
+              File logFile = composeLogToFile.get().asFile
               logger.debug "Logging docker-compose up to: $logFile"
               logFile.parentFile.mkdirs()
               composeLog = new FileOutputStream(logFile)
             }
-            settings.composeExecutor.executeWithCustomOutputWithExitValue(composeLog, args)
-            def servicesToLoad = settings.composeExecutor.getServiceNames()
+            composeExecutor.get().executeWithCustomOutputWithExitValue(composeLog, args)
+            def servicesToLoad = composeExecutor.get().getServiceNames()
             servicesInfos = loadServicesInfo(servicesToLoad).collectEntries { [(it.name): (it)] }
             startCapturing()
             waitForHealthyContainers(servicesInfos.values())
-            if (settings.waitForTcpPorts.get()) {
+            if (waitForTcpPorts.get()) {
                 servicesInfos = waitForOpenTcpPorts(servicesInfos.values()).collectEntries { [(it.name): (it)] }
             }
             printExposedPorts()
-            if (!settings.stopContainers.get()) {
-                settings.serviceInfoCache.set(servicesInfos, getStateForCache())
+            if (!stopContainers.get()) {
+                serviceInfoCache.get().set(servicesInfos, getStateForCache())
             } else {
-                settings.serviceInfoCache.clear()
+                serviceInfoCache.get().clear()
             }
         }
         catch (Exception e) {
             logger.debug("Failed to start-up Docker containers", e)
-            if (!settings.retainContainersOnStartupFailure.get()) {
-                settings.downForcedTask.get().down()
```

Comment:
I'm afraid we cannot just drop this feature :( This is crucial because if there is something wrong in `docker-compose.yml` then we want to `down` the corrupted state, in order to be ready for a next try.

Isn't there any other way how to execute a Gradle task just if some other task failed? 🙏

Reply:
Yes, please check [this](https://github.com/avast/gradle-docker-compose-plugin/pull/351/commits/8c83164c7d3a7db35d6ea268e4e5924cb78e93f2#diff-de1f39ee1c0309d4f56bdaa883c35927bd237df3118f3adab76ce4732fb5f4e6R80) task, it's already implemented :)

Reply:
If I understand it correctly, `finalizedBy` is executed always but we need to execute just in case of failure (the original call was from `catch`). Is this correct? 🙏

Reply:
If this would be problematic to implement, feel free to remove this _convenient_ feature (I actually don't think it is _crucial_).

Reply:
Yes, `finalizedBy` tells the Gradle that it should try to execute `composeDownForcedOnFailure` after each execution of `composeUp`, but before execution Gradle should check if the predicate from `onlyIf` returns `true`. The predicate returns the value of the `startupFailed` property of `ServiceInfoCache` that is set to `true` only when we fall into `catch` block. Since build service is reinstantiated on every build, the initial value is always `false`